---
title: "scatterplot_with_ClinVar"
author: "Tram Nguyen"
date: "2024-04-16"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Function 1: Given a protein, load ClinVar supplement and AM databases.

```{r}
# Open the DuckDB connection
library(AlphaMissenseR)

db <- db_connect()

# From Cheng et al. supplemental data table 5
clinvar <- readRDS("~/AlphaMissense_desktop/data/ClinVar_data.rds")

# eventually included in package as rds:
# data(ClinVar)
```


The ClinVar data gives a transcript ID and a UniProt accession number in the 
protein_variant field. Match these UniProt with our accession number and plot.


```{r}
am_data
am_score
cv_data
cv_score

plot_clinvar <- function(protein, am_data, cv_data, am_score, cv_score){
    
    # join_ClinVar and AlphaMissense
    
    protein <- as.character(protein)
    
    c.cv <- clinvar |>
          filter(accession == protein)
    
    # Check if protein found in ClinVar
    if (nrow(c.cv) < 1) 
      stop(c("No ClinVar information found for the protein accession. ", 
              "Check that the UniProt ID is correct."))
      
    # Grab AM proteins that match CV
    c.AM <- tbl(db, "aa_substitutions") |>
          filter(uniprot_id == protein) |>
          dplyr::as_tibble()
    
    # Check if protein found in AM
    if (nrow(c.AM) < 1) 
      stop(c("No AlphaMissense information found for the protein accession. ", 
              "Check that the UniProt ID is correct."))
      
    # Join databases by protein variant
    res <- as.numeric(gsub(".*?([0-9]+).*", "\\1", c.AM$protein_variant))             
    c.AM <- c.AM %>% mutate(aa_pos = res)
  
    c.combo <- left_join(c.AM, c.cv, by = "protein_variant") |> 
        select(-c('accession', 'AlphaMissense')) |> 
        rename(cv_variant_id = variant_id, cv_label = label) |> 
        relocate('transcript_id', .after = 'uniprot_id')
    
    # Check if protein has multiple transcripts
    if (length(unique(c.combo$transcript_id[!is.na(c.combo$transcript_id)])) > 1) 
      stop(c("Multiple transcripts detected."))
    
    c.combo$transcript_id <- 
      rep(unique(c.combo$transcript_id[!is.na(c.combo$transcript_id)]))
    
    ##### PLOTTING #####
    
    # Grab the thresholds for pathogenicity/benign to plot (dashed lines)
    am_cutoff <- c.combo |> 
                      filter(c.combo$am_class == "ambiguous") |> 
                      select(am_pathogenicity) |> 
                      summarise(path_cutoff = min(am_pathogenicity), 
                                benign_cutoff = max(am_pathogenicity))
    
    
    # Add color code matching AM and CV labels
    c.combo <- c.combo |> 
        mutate(code_color = case_when(
              !is.na(cv_label) & cv_label == "0" ~ "CV_benign",
              !is.na(cv_label) & cv_label == "1" ~ "CV_pathogenic",
              is.na(cv_label) & am_class == "pathogenic" ~ "AM_pathogenic",
              is.na(cv_label) & am_class == "benign" ~ "AM_benign",
              is.na(cv_label) & am_class == "ambiguous" ~ "AM_ambiguous" ))
    
    if (plot_type == "sequence") {
    # Plot sequence window
      
    # Plot sequence window
    library(ggplot2)
    
    plot <- ggplot(c.combo %>% arrange(code_color), aes(aa_pos, am_pathogenicity)) +
        geom_point(aes(shape=code_color, color=code_color, size=code_color, 
                       fill=code_color, stroke=code_color))+
        scale_shape_manual(values=c(19, 19, 19, 21, 21)) +
        scale_discrete_manual(
                aesthetics = "stroke", 
                values = c(0, 0, 0, 1.5, 1.5)) +
        scale_size_manual(values=c(2,2,2,4,4)) +
        scale_color_manual(values=c("gray", "#89d5f5", "#f56c6c", "black", "black")) +
        scale_fill_manual(values=c("gray", "#89d5f5", "#f56c6c", "#007cb0", "#c70606")) +
        geom_hline(yintercept = am_cutoff$path_cutoff, linetype=2, ) +
        geom_hline(yintercept = am_cutoff$benign_cutoff, linetype=2) +
        labs(title = paste0("UniProt ID: ", protein)) +
        xlab("amino acid position") + ylab("AlphaMissense score") +
        theme_classic()


    plot <- plot + theme(
        axis.text.x = element_text(size=16),
        axis.text.y = element_text(size=16),
        axis.title.y = element_text(size=16),
        axis.title.x = element_text(size=16),
        legend.position="none"
        )
    
    results <- plot

  } else if (plot_type == "structure") {
    
    
    # Get AlphaFold structure
    library(r3dmol) 
    
    prediction <- af_predictions(protein)
    glimpse(prediction)

    # Plot AlphaFold-predicted structure
    AF_plot <- c.combo |>
        dplyr::select(-aa_pos) %>% # remove the aa pos column to return to original dataset
        filter(uniprot_id == protein) %>% 
        af_prediction_view()
    
    results <- AF_plot

  } else {
    
    # Invalid option
    results <- "Invalid plot option"
    
  }
  
    results 
    
}

```

# Plotting functionality test

```{r}
plot_clinvar(protein = "P37023", plot_type = "structure") 

```


```{r}
plot_clinvar(protein = "P37023", plot_type = "sequence")

# arg: provide an alphamissense table
# arg: from this previous table, which column stored the AM score you want to use?
# arg: CV data that user can upload
# arg: which column of CV has the scores you want?

# In the man document: give specific details of what you expect the user to load in.
# Otherwise, default is martin's duckDB AM 
```

